File: python/A_star.py
from collections import defaultdict, deque
import heapdict


def printpath(parent,src,goal):
    print(parent)
    l=[goal]
    while src!=goal:
        goal=parent[goal]
        l.append(goal)
    l.reverse()
    print("Path :",l)

def a_star_search(adj,heuristics,src,goal):
    
    open=heapdict.heapdict()
    close=[]
        
    dist={src:0}
    open[src]=heuristics[src]+dist[src]
    parent={src:-1}
    
    while open:
        curr_node,curr_f=open.popitem()
        
        print(curr_node, dist[curr_node],curr_f)
        
        if curr_node==goal:
            print("Solution found")
            printpath(parent, src, goal)
            return 
        
        close.append(curr_node)
        
        for node in adj[curr_node]:
    
            if node[0] not in close:
                
                if node[0] in list(open.keys()):
                    if dist[node[0]]>dist[curr_node]+node[1]:
                        dist[node[0]]=dist[curr_node]+node[1]
                        open[node[0]]=heuristics[node[0]]+dist[node[0]]
                        parent[node[0]]=curr_node
                else:
                    dist[node[0]]=dist[curr_node]+node[1]
                    open[node[0]]=heuristics[node[0]]+dist[node[0]]
                    parent[node[0]]=curr_node
                        
    print("Solution not found")


n=int(input("Enter the number of nodes : "))

nodes=[]
heuristics=dict()

print("Enter",n,"nodes")
for i in range(n):
    x=input("node"+str(i+1)+" : ")
    h=int(input("heuristics value of node "+str(i+1)+" : "))
    nodes.append(x)
    heuristics[x]=h
    
    
print(heuristics)
    
e=int(input("Enter the number of edges : "))

adj=defaultdict(list)

for i in range(e):
    print("Edge",i+1," : ")
    v=input("Start Node : ")
    u=input("End Node : ")
    c=int(input("Enter Cost : "))
    
    if u not in nodes or v not in nodes:
        print("Invalid Edge")
        quit()
    
    adj[v].append((u,c))
    adj[u].append((v,c))
    
#print(adj)
src=input("Enter the source node : ")
goal=input("Enter the goal node : ")


a_star_search(adj,heuristics,src,goal)
    


'''
7
0
14
1
12
2
11
3
6
4
4
5
11
6
0
9
0
1
4
0
2
3
2
3
7
2
4
10
3
4
2
1
4
12
1
5
5
4
6
5
5
6
16

'''    



    

File: informed search/a_star.cpp
#include<bits/stdc++.h>
using namespace std;

vector<int> adj[1000],cost[1000];
int h[1000],f[1000],g[1000];
int p[1000];
int n,e;



void path(int u,int s)
{
    if(u==s)
    {
        cout<<u<<" ";
        return;
    }


    path(p[u],s);
    cout<<u<<" ";
}


void astar(int src,int dest)
{

    for(int i=0; i<n; i++)
    {
        g[i]=INT_MAX;
        f[i]=INT_MAX;
    }
    g[src]=0;
    f[src]=g[src]+h[src];


    priority_queue< pair<int,int> , vector<pair<int,int> >, greater<pair<int,int> > > pq;   //first f(n) , second node

    pq.push({f[src],src});



    while(!pq.empty())
    {
        auto x=pq.top();

        int curr=x.second, currF=x.first;

        //cout<<currF<<" x "<<curr<<endl;

        pq.pop();

        for(int i=0; i<adj[curr].size(); i++)
        {
            int v=adj[curr][i];
            int c=cost[curr][i];

            if(g[curr]+c+h[v]<f[v])
            {
                p[v]=curr;

                g[v]=g[curr]+c;
                f[v]=g[v]+h[v];

                //cout<<f[v]<<" "<<g[v]<<" "<<h[v]<<" * "<<v<<endl;

                pq.push({f[v],v});

            }
        }
    }

    cout<<"Path : ";
    path(dest,src);

    cout<<endl<<f[dest]<<endl;
}

int main()
{
    //cout<<"Enter the number of nodes : ";
    cin>>n;
    //cout<<"Enter the number of edges : ";
    cin>>e;

    //cout<<"Enter "<<e<<" Edges : ";

    int u,v,z;
    for(int i=0; i<e; i++)
    {
        cin>>u>>v>>z;
        adj[u].push_back(v);
        adj[v].push_back(u);
        cost[u].push_back(z);
        cost[v].push_back(z);
    }

    //cout<<"Enter heuristics value of "<<n<<" nodes : "<<endl;
    for(int i=0; i<n; i++)
    {
        cin>>h[i];
    }

    int src,dest;
    cin>>src>>dest;

    astar(src,dest);

}

/*
7 9
0 1 4
0 2 3
1 4 5
1 5 12
2 5 10
2 3 7
3 5 2
4 6 16
5 6 5
14 12 11 6 11 4 0
0 6
*/

/*
Path : 0 2 3 5 6
17
*/


/*
10 14
0 1 6
1 2 3
1 3 2
2 3 1
2 4 5
3 4 8
0 5 3
5 6 1
5 7 7
6 8 3
7 8 2
8 4 5
8 9 3
4 9 5
10 8 5 7 3 6 5 3 1 0
0 9
*/

/*
Path : 0 5 6 8 9
10
*/















/*
8 10
0 1
0 3
0 2
1 4
2 4
2 5
3 5
4 7
5 6
7 6
40 32 25 35 14 17 0 10
0 6
*/

/*
10 9
0 1
0 2
1 3
1 4
2 5
2 6
5 7
6 8
6 9
13 12 4 7 3 8 2 4 9 0
0 9
*/

